deep_qa/data/fields/field.py

```diff
@@ -1,43 +1,53 @@
-"""
-A ``Field`` is some piece of a data instance that ends up as an array in a model (either as an
-input or an output).  Data instances are just collections of fields.
-
-
-Fields go through up to two steps of processing: (1) tokenized fields are converted into token ids,
-(2) fields containing token ids (or any other numeric data) are padded (if necessary) and converted
-into data arrays.  We have two ``Field`` objects corresponding to fields in each stage of this
-processing pipeline: ``UnindexedFields`` are fields containing raw strings (already tokenized) that
-need to be converted into token IDs.  We use these fields to compute a vocabulary, then convert
-them to ``IndexedFields``.  ``IndexedFields`` have a method for determining padding lengths, so
-that a data generator object can intelligently batch together instances, then pad them.
-
-If you are writing a new ``Field`` class, if there is nothing in the field that needs to be
-converted from strings to integers, you should just subclass ``IndexedField``.  If you `do` need to
-convert strings to integers, you should subclass both ``IndexedField`` and ``UnindexedField``, and
-make the ``UnindexedField`` subclass public, and the other private.
-"""
 from typing import Dict, List
 
 import numpy
 
 from ..vocabulary import Vocabulary
 
-# Because of how the dependencies work between these objects, we're defining them in reverse order,
-# so that UnindexedField can reference IndexedField in its return types.
-
-class IndexedField:
+class Field:
     """
-    An ``IndexedField`` is a field that is almost ready to be put into a data array for use in a
-    model.  The only thing missing is padding.  The methods on this object allow you to get padding
-    lengths, so you can group things into batches of similar size, and then actually pad the
-    field into a fixed-size array that can be used in a batch of instances.
+    A ``Field`` is some piece of a data instance that ends up as an array in a model (either as an
+    input or an output).  Data instances are just collections of fields.
+
+    Fields go through up to two steps of processing: (1) tokenized fields are converted into token
+    ids, (2) fields containing token ids (or any other numeric data) are padded (if necessary) and
+    converted into data arrays.  The ``Field`` API has methods around both of these steps,
+    though they may not be needed for some concrete ``Field`` classes - if your field doesn't have
+    any strings that need indexing, you don't need to implement ``count_vocab_items`` or ``index``.
+    These methods ``pass`` by default.
+
+    Once a vocabulary is computed and all fields are indexed, we will determine padding lengths,
+    then intelligently batch together instances and pad them into actual arrays.
     """
+    def count_vocab_items(self, counter: Dict[str, Dict[str, int]]):
```

Comment:
Different fields may or may not use the same namespace.  Text fields will by default all use the same namespace (so both sentences get the same token ids), but you could change that if you want, e.g., to give different ids for characters in a structured name field, if you really wanted to.

We also use this same mechanism to map tags and labels to integers (so, e.g., you don't need to do your own mapping from "B-PERS" to integers), and those will use a different namespace by default.

Regarding pre-trained vectors, if I have a word in my training data that appears often, but is not in my pre-trained vectors, I still want to assign it a non-OOV id, and learn a vector for it.  Or if I'm doing Turkish NLP, or using byte-level character representations, I might not have pre-trained vectors at all.  We definitely want to be able to count vocab items that occur in the data, and be able to map frequently occurring items in the data to unique ids.

I agree that we _also_ might want to give ids to any token I have a pre-trained vector for.  That's not incompatible with this - we're just computing counts here, not constructing the vocabulary, and the code in `Trainer` that constructs the vocabulary could add all tokens to the vocab that have a pre-trained vector.  That's not hard.